assertj-core/src/test/java/org/assertj/core/api/date/DateAssert_isEqualTo_Test.java

```diff
@@ -12,42 +12,125 @@
+
+    Arguments[] should_pass() {
+      return new Arguments[] {
+          arguments(Date.from(parse("1970-01-01T00:00:00.000000001Z")),
+                    Date.from(parse("1970-01-01T00:00:00.000000001Z"))),
+          arguments(Date.from(parse("1970-01-01T00:00:00.000000001Z")),
+                    Timestamp.from(parse("1970-01-01T00:00:00.000000001Z")))
+      };
+    }
+
+    @ParameterizedTest
+    @MethodSource
+    void should_fail(Date actual, Object expected) {
+      // WHEN
+      AssertionError error = expectAssertionError(() -> assertThat(actual).isEqualTo(expected));
+      // THEN
+      then(error).isInstanceOf(AssertionFailedError.class);
+    }
+
+    Arguments[] should_fail() {
+      return new Arguments[] {
+          arguments(Timestamp.from(parse("1970-01-01T00:00:00.000000001Z")),
+                    Date.from(parse("1970-01-01T00:00:00.000000001Z")))
+      };
+    }
 
-  protected DateAssert assertionInvocationWithInstantArg(Instant instant) {
-    return assertions.isEqualTo(instant);
   }
 
-  @Override
-  protected void verifyAssertionInvocation(Date date) {
-    verify(objects).assertEqual(getInfo(assertions), getActual(assertions), date);
+  @Nested
+  @TestInstance(PER_CLASS)
+  class With_Instant {
+
+    @ParameterizedTest
+    @MethodSource
+    void should_pass(Date actual, Instant expected) {
+      // WHEN/THEN
+      assertThat(actual).isEqualTo(expected);
+    }
+
+    Arguments[] should_pass() {
+      return new Arguments[] {
+          arguments(Date.from(parse("1970-01-01T00:00:00.000000001Z")),
+                    parse("1970-01-01T00:00:00.000000001Z")),
+          arguments(Timestamp.from(parse("1970-01-01T00:00:00.000000001Z")),
+                    parse("1970-01-01T00:00:00.000000001Z"))
+      };
+    }
+
   }
 
-  @Override
-  protected DateAssert assertionInvocationWithInstantArg() {
-    return assertions.isEqualTo(otherDate.toInstant());
+  @Nested
+  @TestInstance(PER_CLASS)
+  class With_String {
+
+    @ParameterizedTest
+    @MethodSource
+    void should_pass(Date actual, String expected) {
+      // WHEN/THEN
+      assertThat(actual).isEqualTo(expected);
+    }
+
+    Arguments[] should_pass() {
+      return new Arguments[] {
+          arguments(Date.from(parse("1970-01-01T00:00:00.000000001Z")),
+                    "1970-01-01T00:00:00.000Z")
+      };
+    }
+
+    @ParameterizedTest
+    @MethodSource
+    void should_fail(Date actual, Object expected) {
+      // WHEN
+      AssertionError error = expectAssertionError(() -> assertThat(actual).isEqualTo(expected));
+      // THEN
+      then(error).isInstanceOf(AssertionFailedError.class);
+    }
+
+    Arguments[] should_fail() {
+      return new Arguments[] {
+          arguments(Date.from(parse("1970-01-01T00:00:00.000000001Z")),
+                    "1970-01-01T00:00:00.000000001Z"),
+          arguments(Timestamp.from(parse("1970-01-01T00:00:00.000000001Z")),
+                    "1970-01-01T00:00:00.000000001Z")
```

Comment:
The fact that these cases currently belong to failures might not be 100% correct.

However, I don't see an easy way forward right now so I'd rather document the current behavior with them and we can address this topic in a second step.

Reply:
I think we are hitting java `Date` limitation as `SimpleDateFormat` does not handle nanos.
https://stackoverflow.com/questions/30135025/java-date-parsing-with-microsecond-or-nanosecond-accuracy

The best course of action is to document that and point to not to use date assertion accepting string, which was a handy shortcut for users at the time to easily create dates but now with the java.time API it is not necessary and I would even say something we could deprecate.

Reply:
I wonder if cases like this one and #3382 should be addressed differently, for example dropping the internal usage of `SimpleDateFormat` and favoring `DateTimeFormatter` instead.

For now, I propose to merge these changes as they are, for sure they don't make anything worse, and we can follow up on the `SimpleDateFormat` topic separately.

WDYT?

Reply:
Let's merge this and use `DateTimeFormatter` as it handles nanoseconds

Reply:
#3482 raised for tracking.